{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"name":"ORFfinder_R.ipynb","provenance":[],"collapsed_sections":[],"toc_visible":true},"kernelspec":{"display_name":"R","name":"ir"}},"cells":[{"cell_type":"markdown","metadata":{"id":"UrdIRDY7NEkT"},"source":["ORF Finding Method\n","\n","The ORF finder is primarily composed of three functions\n","\n","*   codons(seq)\n","*   ORF(condonPositions)\n","*   ORFfinder(dta, nucleotides, sel_gen, rev)\n","\n","codons() will find all positions of start and stop codons and place them into three lists, one for each frame. Once found, the ORF() function will accept these lists and validate/pair the start and stop codons based on a set minimum length as well as where the previous ORF was found. For each valid start codon, multiple valid stop codons are paired in order to obtain 3 ORFs. Both of these functions are called repeatedly within the ORFfinder() function. Within this function, the ORFcheck() function also runs to determine how many genes the ORF finder finds and how many extraneous ORFs it finds."]},{"cell_type":"code","metadata":{"id":"uJcF9kXNGFJq"},"source":["## Upload data from Module 6 data folder: \n","\n","# all_gene_positions.txt\n","# all_incorrect_ORFS.csv\n","# full_sequence.fasta\n","# genome1.fasta\n","# genome1_gene_positions.txt\n","# genome2.fasta\n","# rev_comp.fasta"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"q2y3YodnaNO1"},"source":["<font color='blue'>Cell 1 Installing and loading necessary packages"]},{"cell_type":"code","metadata":{"id":"4IBrlleqHhAc"},"source":["#install.packages(\"dplyr\")\n","library(\"dplyr\") #The dplyr package is commonly employed for dataframe manipulation"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"FgRigBJsz7s3"},"source":["<font color='blue'>Cell 2 \n","Reading data"]},{"cell_type":"code","metadata":{"id":"4CKPjIYQ6zmo"},"source":["#Import gene position data \n","#The first element of each row is the start position of the gene\n","#The second is the end position of the gene\n","all_gene_data <- read.table(\"all_gene_positions.txt\", header=FALSE, sep=\",\", quote=\"\") \n","gene_position <- as.data.frame(all_gene_data[1:2269,]) #Data for forward direction\n","gene_position_r <- as.data.frame(all_gene_data[2270:4584,]) #Data for reverse complement\n","\n","#Read full forward direction genome\n","full_seq <- read.table(\"full_sequence.fasta\", header=FALSE, sep=\"\\n\", quote=\"\", stringsAsFactors=FALSE) \n","full_seq <- as.data.frame(full_seq[-1,], stringsAsFactors=FALSE) #Remove fasta descriptor (first row)\n","full_seq <- paste(full_seq[,1], sep=\"\\n\", collapse=\"\") #Turn data into one character string by removing line returns\n","\n","#Read full reverse complement direction genome\n","full_rev <- read.table(\"rev_comp.fasta\", header=FALSE, sep=\"\\n\", quote=\"\", stringsAsFactors=FALSE) \n","full_rev <- as.data.frame(full_rev[-1,], stringsAsFactors=FALSE) #Remove fasta descriptor (first row)\n","full_rev <- paste(full_rev[,1], sep=\"\\n\", collapse=\"\") #Turn data into one character string by removing line returns"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"WTOYpjbqEs8y"},"source":["\n","\n","# ORF Finder"]},{"cell_type":"markdown","metadata":{"id":"g21kQsiE0Act"},"source":["<font color='blue'>Cell 3\n","Finding codon positions"]},{"cell_type":"markdown","metadata":{"id":"3C3LC3yZ1bGq"},"source":["codons(): This function will find the positions at which a start codon or stop codon appears. This function will return two lists (starts and stops), each with three sublists that contain the found positions in three frames.\n","\n","Inputs: \n","-   seq: The sequence being considered, ie. forward or reverse complement sequences (full_seq and full_rev, respectively)\n","-   start_c: The start codon considered. This function is called from another function that iterates through ATG, GTG, TTG\n","\n","Inside the function:\n","-   stops: The stop codons that will be found\n","-   lst1: A list that will contain the start codon positions\n","-   lst2: A list that will contain the stop codon positions\n","      - Both lst1 and lst2 are a list of 3 lists where each sublist is the list of positions for each frame.\n","\n","Output:\n","Returns the list codonPositions which contains lst1 and lst2 "]},{"cell_type":"code","metadata":{"id":"ymU7_58vZqyu"},"source":["#ORF Finder part 1: Find codon positions\n","codons <- function(seq, start_c) {\n","  \n","  stops <- c(\"TAA\",\"TGA\",\"TAG\")\n","  codons <- list() #Empty list that will hold nucleotide triplets (codons)\n","  lst1 <- list() #Empty list that will hold 3 lists of start positions, one list for each reading frame\n","  lst2 <- list() #Empty list that will hold 3 lists of stop positions, one list for each reading frame\n","  \n","  #Split sequence into reading frames using the sub() function, which removes unread nucleotides at the start of the sequence\n","  frame1 <- seq #First frame starts reading at first nucleotide\n","  frame2 <- sub(\".\", \"\", seq) #Second frame starts reading at second nucleotide\n","  frame3 <- sub(\"..\", \"\", seq) #Third frame strats reading at thrid nucleotide\n","  frames <- list(frame1, frame2, frame3) #Create list of 3 reading frames\n","  \n","  #Cut frames into triplets and store in the list codons\n","  #The functions gregexpr() and regmatches() find and extract the triplets, respectively\n","  for (i in 1:length(frames)){ \n","    codons[[i]] <- regmatches(frames[[i]], gregexpr(\".{3}\", frames[[i]]))[[1]]\n","  }\n","  \n","  #Postion of all start codons from all frames\n","  #Start by producing lists of the positions of each start codon in each frame using the grep() function\n","  for (i in 1:length(start_c)) { \n","    lst1[[i]] <- list(3*grep(pattern=start_c[i], codons[[1]])-2, 3*grep(pattern=start_c[i], codons[[2]])-1, 3*grep(pattern=start_c[i], codons[[3]])) #Multiply the codon position by 3 and adjust for reading frame to obtain the nucleotide position of each start codon\n","  }\n","  #Then combine start codon lists to obtain one list of all start codons for each frame using the mapply() function\n","  lst1 <- mapply(c, lst1[[1]], lst1[[2]], lst1[[3]]) \n","  \n","  #Repeat above for stop codons\n","  for (i in 1:length(stops)) {\n","    lst2[[i]] <- list(3*grep(pattern=stops[i], codons[[1]])-2, 3*grep(pattern=stops[i], codons[[2]])-1, 3*grep(pattern=stops[i], codons[[3]]))\n","  }\n","  lst2 <- mapply(c, lst2[[1]], lst2[[2]], lst2[[3]])\n","  \n","  #Finally return lst1 and lst2 in a list called codonPositions. We must do this because R functions can only return one element.\n","  codonPositions <- list(lst1, lst2)\n","  return(codonPositions)\n","  \n","}"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"QMOUQ_6H0BbL"},"source":["<font color='blue'>Cell 4\n","Determining valid ORFs"]},{"cell_type":"markdown","metadata":{"id":"3jYongTi1qh6"},"source":["ORF(): Pairs the start and stop codon positions found with the codons function to form valid ORF start-stop pairs.\n","\n","Inputs:\n","*   codonPositions is the output of the codons() function\n","*   min_len is the minimum length used for determining valid ORFs\n","\n","Inside the function:\n","*   ORFstarts and ORFstops: Lists that will contain the starts and stops of the determined valid ORFs\n","*   all_frames: When looking for ORFs, all start codon positions for all frames will be put into one list. This allows to scan for valid ORF starts sequentially across all frames.\n","*   min_len: The minimum ORF length is 40\n","*   start_min: Starting from zero, this variable is updated to dictate the position from which the next ORF start will be considered. After a start codon has been paired with its stop codons, this variable is updated with the min_len and the look_back.\n","*   stop_count: The stop position for loop iterates until each valid start codon is paired with 3 stops so long as the min_len condition is met. \n","*   look_back: This parameter is subtracted from the start_min variable to account for the possibility of overlapping ORFs\n","\n","Output: \n","-   ORFs: a list that contains ORFstarts, ORFstops and found_frame. ORFstarts and ORFstops contain the start and stop positions of codons that form true ORFs. Each element in the ORFstarts list is paired with the corresponding element of the ORFstops list, eg. ORFstarts[1] and ORFstops[1] are the indices of the first ORF that was found by the ORF() function. The third vector, found_frame, contains the reading frames for all valid start codons (only used for diagnostic purposes).\n"]},{"cell_type":"code","metadata":{"id":"_rYCbzQw1vE8"},"source":["#ORF Finder part 2: Determine valid ORFs\n","ORF <- function(codonPositions, min_len) {\n","  \n","  ORFstarts <- as.numeric() #Empty vector for start positions\n","  ORFstops <- as.numeric() #Empty vector for stop positions\n","  found_frame <- as.numeric() #Empty vector for reading frames of all valid start codons\n","  look_back <- 5 #To account for overlapping ORFs\n","  start_min <- 0 #Set minimum start position\n","  #min_len <- 40\n","  \n","  #Consolidate and sort start and stop codon positions in ascending order\n","  lst1 <- lapply(codonPositions[[1]], sort) #Maintain 3 lists for reading frames, each list in ascending order\n","  lst2 <- lapply(codonPositions[[2]], sort) #Maintain 3 lists for reading frames, each list in ascending order\n","  all_frames <- sort(unlist(lst1)) #Consolidate start codon positions into one vector, in ascending order\n","\n","  for (s_cod in all_frames) { \n","        \n","    for (s_frame in 1:length(lst1)) {  #Find the frame of the start codon\n","      if (s_cod %in% lst1[[s_frame]]) {\n","        frame <- s_frame\n","        found_frame <- append(found_frame, frame) #Add the frame to the vector found_frame\n","        }\n","      }\n","      \n","    stop_count <- 1 #Set number of stop positions to begin at 1 for each start codon tested \n","\n","    for (stp_cod in lst2[[frame]]) {\n","      if (s_cod > start_min) { #Start looking for stops in the same frame as the start codon\n","        if (stop_count > 3) { #The minimum start position will be reset after finding 3 stop codons\n","          start_min <- start_min + min_len - look_back #Start position will be less than the new minimum start position, so the stop position \"for\" loop will break and the next start codon will be tested\n","        }\n","          \n","        else {\n","          if (stp_cod > s_cod + min_len) { #Before breaking out of the stop position \"for\" loop, add stop position to accompany the start position in corresponding lists ORFstops and ORFstarts \n","            ORFstarts <- append(ORFstarts, s_cod)\n","            ORFstops <- append(ORFstops, stp_cod)\n","            stop_count <- stop_count + 1 #Move to next corresponding stop codon\n","          }\n","        }\n","      }\n","    }\n","  }\n","\n","  ORFs <- list(ORFstarts, ORFstops, found_frame)\n","  return(ORFs)\n","\n","}"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"HE4vLTs10CnQ"},"source":["<font color='blue'>Cell 5\n","ORF checking function"]},{"cell_type":"markdown","metadata":{"id":"puDjG24ARuek"},"source":["ORFcheck(): Checks which of the ORF start-stop pairs are coding (ie. found in the gene position data) and which are non coding (ie. not found in the gene position data).\n","\n","Inputs:\n","*   ORFs from the ORF() function\n","*   gene_pos: The gene position data (either gene_position or gene_position_r)\n","*   rev: A TRUE/FALSE parameter that is set to TRUE if looking at the reverse complement (with gene_position_r), and FALSE otherwise\n","\n","Inside the function:\n","-   OC: start-stop pair being tested\n","*   corr: The number of found ORFs that are contained in the gene position data (used for diagnostic purposes)\n","*   corORF: A dataframe of the ORFs (start and end positions) that are contained in the gene position data\n","*   incorORF: A dataframe of the ORFs (start and end positions) that are not contained in the gene position data (non coding ORFs)\n","\n","Output: Returns ORFcheckData, a list containing corr, corORF and incorORF."]},{"cell_type":"code","metadata":{"id":"mmcg3zMeB6GD"},"source":["#ORF Finder part 3: Determine which found ORFs are coding and non coding\n","ORFcheck <- function(ORFs, gene_pos, rev) {\n","  \n","  ORFstarts <- ORFs[[1]] \n","  ORFstops <- ORFs[[2]] \n","\n","  if (rev == TRUE) {\n","    ORFfound <- data.frame(ORFstarts-1, ORFstops+2) #Different indexing than forward genome. Add 2 to stop position to obtain position of last nucleotide in stop codon.\n","  }\n","  \n","  else {\n","    ORFfound <- data.frame(ORFstarts, ORFstops+2) #Add 2 to stop position to obtain position of last nucleotide in stop codon.\n","  }  \n","  \n","  colnames(ORFfound) <- c(\"ORFstarts\", \"ORFstops\") #Renaming the columns will enable us to use the dplyr join functionalities to find the coding and non coding ORFs.\n","  colnames(gene_pos) <- c(\"ORFstarts\", \"ORFstops\")\n","  corORF <- inner_join(ORFfound, gene_pos, by=c(\"ORFstarts\", \"ORFstops\")) #The inner join yields all ORFs that were found and coding (ie. found in the gene position data)\n","  incorORF <- anti_join(ORFfound, gene_pos, by=c(\"ORFstarts\", \"ORFstops\")) #The anti join yields all ORFs that were found and non coding (ie. not found in the gene position data)\n","  corr <- nrow(corORF) #Count the number of correct ORFs\n","  ORFcheckData <- list(corr, corORF, incorORF)\n","  return(ORFcheckData)\n","  \n","}"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"KTEKx-WE0Dh1"},"source":["<font color='blue'>Cell 6\n","Running the previous functions"]},{"cell_type":"markdown","metadata":{"id":"zSaEja9sW4ld"},"source":["ORFfinder(): This function simply runs the functions defined above and returns the dataframes of all coding and non coding ORFs found\n","\n","Inputs:\n","*   dta: Either full_seq or full_rev (forward genome sequence or reverse complement genome sequence)\n","*   nucleotides: A parameter to limit the ORFfinder search to reduce processing time (ie. nucleotides = 10000 will only look at the first 10000 nucleotides in the genome)\n","*   sel_gen: The gene position data that is also limited by the 'nucleotides' parameter\n","*  rev and min_len as defined in previous functions\n","\n","Outputs: Returns the list ORFfinderData which is a lists composed of two dataframes, one that holds the start-stop pairs for coding ORFs (corORF) and one that holds the start-stop pairs for non coding ORFs (incorORF)"]},{"cell_type":"code","metadata":{"id":"jgIu1npIXSMj"},"source":["#ORF Finder part 4: Putting all functions together\n","ORFfinder <- function(dta, nucleotides, sel_gen, rev, min_len) {\n","\n","  start_codons <- c(\"ATG\",\"GTG\",\"TTG\")\n","  codonPositions <- codons(dta[1:nucleotides], start_codons)\n","  ORFs <- ORF(codonPositions, min_len)\n","  ORFcheckData <- ORFcheck(ORFs, sel_gen, rev)\n","  ORFfinderData <- list(ORFcheckData[[2]], ORFcheckData[[3]]) #Note ORFcheckData[[2]] is the dataframe corORF (coding ORFs), while ORFcheckData[[3]] is the dataframe incorORF (non coding ORFs)\n","  return(ORFfinderData)\n","  \n","}"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"U2xJttonPnCp"},"source":["# ORF Finder with an E. coli genome\n","\n"]},{"cell_type":"markdown","metadata":{"id":"kYKJhu6aVztz"},"source":["<font color='blue'>Cell 7\n","Finding ORFs on a section of the E. Coli genome"]},{"cell_type":"markdown","metadata":{"id":"5VYqT3PnYCu-"},"source":["genome_section(): This function takes inputs from the user and runs the ORFfinder() function. This function prints how many coding ORFs and non coding ORFs were found.\n","\n","Output: Returns a list called all_pred which contains 2 dataframes, one that contains all start-stop pairs for coding and non coding ORFs one the forward genome (all_f_pred), and one that contains all start-stop pairs for coding and non coding ORFs one the reverse complement genome (all_r_pred)."]},{"cell_type":"code","metadata":{"id":"4VeWq95VYbhl"},"source":["genome_section <- function() {\n","  #Get the necessary parameters from the user\n","  nucleotides <- readline(prompt=\"Please enter the number of nucleotides to analyze: \")\n","  nucleotides <- as.numeric(nucleotides) #Convert from a string to a numeric\n","  min_len <- readline(prompt=\"Please enter the minimum length of ORFs to consider: \")\n","  min_len <- as.numeric(min_len)\n","  \n","  #Limit the forward and reverse complement using the 'nucleotides' parameter. The substr() function subsets the genome to obtain nucleotides with indices 1 to the nucleotides threshold.\n","  forward_section <- substr(full_seq, start=1, stop=nucleotides)\n","  reverse_section <- substr(full_rev, start=1, stop=nucleotides)\n","  \n","  #Limiting the gene position data by the 'nucleotides' parameter \n","  #The apply() function assigns TRUE to rows (row is indicated by the 1 argument of the apply() function) when any value in the row is greater than the 'nucleotides' parameter, and FALSE otherwise. By using the logical NOT operator (!) before the apply() function, we are limiting the gene position data to ORFs that are found within our nucleotide threshold.\n","  selected_genes_f <- gene_position[!apply(gene_position, 1, function(x) any(x > nucleotides)),] \n","  selected_genes_r <- gene_position_r[!apply(gene_position_r, 1, function(x) any(x > nucleotides)),]\n","\n","  #Running the ORFfinder() function on the forward and reverse complement with the user entered parameters\n","  f_ORFfinderData <- ORFfinder(dta=forward_section, nucleotides=nucleotides, sel_gen=selected_genes_f, rev=FALSE, min_len=min_len)\n","  r_ORFfinderData <- ORFfinder(dta=reverse_section, nucleotides=nucleotides, sel_gen=selected_genes_r, rev=TRUE, min_len=min_len)\n","  \n","  #Consolidate results\n","  #Note f_ORFfinderData[[1]] is a dataframe of coding ORFs, while f_ORFfinderData[[2]] is a dataframe for non coding ORFs\n","  all_f_pred <- list(f_ORFfinderData[[1]], f_ORFfinderData[[2]])\n","  all_r_pred <- list(r_ORFfinderData[[1]], r_ORFfinderData[[2]])\n","  all_pred <- list(all_f_pred, all_r_pred)\n","\n","  #return(all_pred)\n","\n","  #Printing the number of genes found within the nucleotides threshold, the number of coding ORFs and non coding ORFs found using the ORFfinder() function\n","  cat(paste(\"-----------------------------------------------------------\", \n","    paste0(\"There are \", nrow(selected_genes_f), \" genes on the forward strand\"), \n","    paste0(\"There are \", nrow(selected_genes_r), \" genes on the reverse strand\"),\n","    \"-----------------------------------------------------------\",\n","    \"Forward:\", \n","    paste0(length(f_ORFfinderData[[1]][[1]]), \" found correctly\"), \n","    paste0(length(f_ORFfinderData[[2]][[1]]), \" additional ORFs found\"),\n","    \"-----------------------------------------------------------\",\n","    \"Reverse:\", \n","    paste0(length(r_ORFfinderData[[1]][[1]]), \" found correctly\"), \n","    paste0(length(r_ORFfinderData[[2]][[1]]), \" additional ORFs found\"),\n","    sep=\"\\n\")) #This is one technique for printing text as the output of a function\n","\n","}\n","\n","genome_section()"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"dsmE67gJV7pE"},"source":["<font color='blue'>Cell 8\n","Running the previous functions on a fraction of the E. coli genome"]},{"cell_type":"markdown","metadata":{"id":"J4jESR8xiHXD"},"source":["This code block outputs the number of ORFs found along with tables of the first 5 ORFs for both the forward and the reverse complement genome sequences. It also contains two functions, test_genome_ORFs (), which obtains a fraction of the imported genome sequence for ORF testing, and rev_comp(), which obtains the reverse complement sequence of the imported genome sequence."]},{"cell_type":"code","metadata":{"id":"c6LeSy9Lh50g"},"source":["genome_filename <- \"genome1.fasta\"\n","\n","#The test_genome_ORFs() function is used to obtain a fraction of the original genome for ORF detection. Use this to reduce processing time.\n","test_genome_ORFs <- function(lim) { #The limit argument will be a float between 0 and 1\n","  test_genome <- as.character(read.table(file=genome_filename, header=FALSE, sep=\"\\n\", quote=\"\", stringsAsFactors=FALSE))\n","  test_genome <- substr(test_genome, start=1, stop=nchar(test_genome)*lim)\n","  return(test_genome)\n","}\n","\n","#Finding reverse complement of genome sequence with the rev_comp() function \n","rev_comp <- function(test_genome) {\n","  test_genome_split <- strsplit(test_genome, \"\")[[1]] #Split genome into single characters\n","  test_genome_reversed <- rev(test_genome_split) #Reverse the split genome\n","  test_gen_rcomp <- character() #Create an empty character vector that will become the reverse complement sequence\n","\n","  for (i in test_genome_reversed) {\n","  \n","    if (i == \"A\") {\n","      j <- \"T\"\n","    }\n","  \n","    else if (i == \"T\") {\n","      j <- \"A\"\n","    }\n","  \n","    else if (i == \"C\") {\n","      j <- \"G\"\n","    }\n","  \n","    else {\n","      j <- \"C\"\n","    }\n","  \n","    test_gen_rcomp <- append(test_gen_rcomp, j)\n","  \n","  }\n","\n","  test_genome_r <- paste(test_gen_rcomp, collapse=\"\") #Consolidate the single characters into one string\n","  return(test_genome_r)\n","\n","}\n","\n","size_limit <- readline(prompt=\"Enter a value between 0 and 1 to limit the genome: \")\n","size_limit <- as.numeric(size_limit)\n","min_len <- readline(prompt=\"Please enter the minimum length of ORFs to consider: \")\n","min_len <- as.numeric(min_len)\n","\n","test_genome <- test_genome_ORFs(size_limit)\n","\n","#Find ORFs for forward genome\n","start_codons <- c(\"ATG\",\"GTG\",\"TTG\")\n","codonPositions <- codons(test_genome, start_codons)\n","ORFs <- ORF(codonPositions, min_len)\n","ORFstarts <- ORFs[[1]]\n","ORFstops <- ORFs[[2]]\n","forward_ORFs <- data.frame(ORFstarts, ORFstops) #Create dataframe to hold the start-stop pairs\n","colnames(forward_ORFs) <- c(\"ORF start location\", \"ORF stop location\") #Rename the columns using the colnames() function\n","\n","#Repeat for reverse compliment genome\n","test_genome_r <- rev_comp(test_genome)\n","codonPositions_r <- codons(test_genome_r, start_codons)\n","ORFs_r <- ORF(codonPositions_r, min_len)\n","ORFstarts_r <- ORFs_r[[1]]\n","ORFstops_r <- ORFs_r[[2]]\n","reverse_ORFs <- data.frame(ORFstarts_r, ORFstops_r)\n","colnames(reverse_ORFs) <- c(\"ORF start location\", \"ORF stop location\")\n","\n","#This is another technique for printing text as the output of a function \n","noquote(\"-----------------------------------------------------------\")\n","noquote(paste0(\"ORFs found on the forward direction:            \", nrow(forward_ORFs)))\n","noquote(paste0(\"ORFs found on the reverse complement direction: \", nrow(reverse_ORFs)))\n","\n","noquote(\"-----------------------------------------------------------\")\n","noquote(\"The first 5 ORFs found on the forward direction: \")\n","noquote(head(forward_ORFs, n=5))\n","\n","noquote(\"-----------------------------------------------------------\")\n","noquote(\"The first 5 ORFs found on the reverse complement direction: \")\n","noquote(head(reverse_ORFs, n=5))\n"],"execution_count":null,"outputs":[]}]}